File: backend/domains/task/model/task.go
package model

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// TaskStatus 定义任务状态
type TaskStatus string

const (
	StatusPending    TaskStatus = "pending"
	StatusInProgress TaskStatus = "in_progress"
	StatusCompleted  TaskStatus = "completed"
)

// IsValid 检查任务状态是否有效
func (s TaskStatus) IsValid() bool {
	switch s {
	case StatusPending, StatusInProgress, StatusCompleted:
		return true
	}
	return false
}

// Priority 定义任务优先级
type Priority string

const (
	PriorityLow    Priority = "low"
	PriorityMedium Priority = "medium"
	PriorityHigh   Priority = "high"
)

// IsValid 检查任务优先级是否有效
func (p Priority) IsValid() bool {
	switch p {
	case PriorityLow, PriorityMedium, PriorityHigh:
		return true
	}
	return false
}

// Tag 任务标签
type Tag struct {
	Name  string
	Color string
}

// Task 聚合根
// 包含任务的所有核心属性和业务行为
type Task struct {
	ID          string
	UserID      string // 所属用户 ID
	Title       string
	Description string
	Status      TaskStatus
	Priority    Priority
	DueDate     *time.Time // 可选
	Tags        []Tag
	CreatedAt   time.Time
	UpdatedAt   time.Time
	CompletedAt *time.Time // 完成时间
}

// 领域错误定义
var (
	ErrTaskTitleEmpty       = fmt.Errorf("TASK_TITLE_EMPTY: 任务标题不能为空")
	ErrTaskAlreadyCompleted = fmt.Errorf("TASK_ALREADY_COMPLETED: 任务已完成")
	ErrInvalidDueDate       = fmt.Errorf("INVALID_DUE_DATE: 截止日期无效")
	ErrTagNameEmpty         = fmt.Errorf("TAG_NAME_EMPTY: 标签名不能为空")
	ErrTooManyTags          = fmt.Errorf("TOO_MANY_TAGS: 标签过多，最多 10 个")
	ErrDuplicateTag         = fmt.Errorf("DUPLICATE_TAG: 标签重复")
	ErrInvalidPriority      = fmt.Errorf("INVALID_PRIORITY: 优先级无效")
)

// NewTask 创建一个新的任务
func NewTask(userID, title, description string, priority Priority) (*Task, error) {
	if userID == "" {
		return nil, fmt.Errorf("USER_ID_REQUIRED: 用户 ID 不能为空")
	}
	if title == "" {
		return nil, ErrTaskTitleEmpty
	}
	if len(title) > 200 {
		return nil, fmt.Errorf("TASK_TITLE_TOO_LONG: 标题过长，最大 200 字符")
	}
	if len(description) > 5000 {
		return nil, fmt.Errorf("TASK_DESCRIPTION_TOO_LONG: 描述过长，最大 5000 字符")
	}
	if !priority.IsValid() {
		return nil, ErrInvalidPriority
	}

	now := time.Now()
	return &Task{
		ID:          uuid.New().String(),
		UserID:      userID,
		Title:       title,
		Description: description,
		Status:      StatusPending,
		Priority:    priority,
		Tags:        []Tag{},
		CreatedAt:   now,
		UpdatedAt:   now,
	}, nil
}

// Update 更新任务信息
func (t *Task) Update(title, description string, priority Priority) error {
	if t.Status == StatusCompleted {
		return ErrTaskAlreadyCompleted
	}
	if title == "" {
		return ErrTaskTitleEmpty
	}
	if len(title) > 200 {
		return fmt.Errorf("TASK_TITLE_TOO_LONG: 标题过长，最大 200 字符")
	}
	if len(description) > 5000 {
		return fmt.Errorf("TASK_DESCRIPTION_TOO_LONG: 描述过长，最大 5000 字符")
	}
	if !priority.IsValid() {
		return ErrInvalidPriority
	}

	t.Title = title
	t.Description = description
	t.Priority = priority
	t.UpdatedAt = time.Now()
	return nil
}

// SetDueDate 设置截止日期
func (t *Task) SetDueDate(dueDate time.Time) error {
	// 验证日期不能早于当前时间
	if dueDate.Before(time.Now()) {
		return ErrInvalidDueDate
	}
	t.DueDate = &dueDate
	t.UpdatedAt = time.Now()
	return nil
}

// Complete 标记任务为已完成
func (t *Task) Complete() error {
	if t.Status == StatusCompleted {
		return ErrTaskAlreadyCompleted // 已经完成，返回错误
	}
	t.Status = StatusCompleted
	now := time.Now()
	t.CompletedAt = &now
	t.UpdatedAt = now
	return nil
}

// AddTag 添加标签
func (t *Task) AddTag(tag Tag) error {
	if tag.Name == "" {
		return ErrTagNameEmpty
	}
	if len(t.Tags) >= 10 {
		return ErrTooManyTags
	}
	// 检查重复
	for _, existing := range t.Tags {
		if existing.Name == tag.Name {
			return ErrDuplicateTag
		}
	}
	t.Tags = append(t.Tags, tag)
	t.UpdatedAt = time.Now()
	return nil
}

// RemoveTag 移除标签
func (t *Task) RemoveTag(tagName string) {
	newTags := []Tag{}
	for _, tag := range t.Tags {
		if tag.Name != tagName {
			newTags = append(newTags, tag)
		}
	}
	t.Tags = newTags
	t.UpdatedAt = time.Now()
}

File: backend-nodejs/domains/task/model/task.ts
/**
 * Task 领域模型
 * 聚合根：包含任务的所有核心属性和业务行为
 */

import { randomUUID } from 'crypto';

// 任务状态
export type TaskStatus = 'pending' | 'in_progress' | 'completed';

export const TaskStatuses = {
  Pending: 'pending' as TaskStatus,
  InProgress: 'in_progress' as TaskStatus,
  Completed: 'completed' as TaskStatus,
} as const;

// 优先级
export type Priority = 'low' | 'medium' | 'high';

export const Priorities = {
  Low: 'low' as Priority,
  Medium: 'medium' as Priority,
  High: 'high' as Priority,
} as const;

// 标签
export interface Tag {
  name: string;
  color: string;
}

// Task 聚合根
export class Task {
  id: string;
  userId: string;
  title: string;
  description: string;
  status: TaskStatus;
  priority: Priority;
  dueDate: Date | null;
  tags: Tag[];
  createdAt: Date;
  updatedAt: Date;
  completedAt: Date | null;

  constructor(
    id: string,
    userId: string,
    title: string,
    description: string,
    priority: Priority,
    status: TaskStatus = TaskStatuses.Pending,
    dueDate: Date | null = null,
    tags: Tag[] = [],
    createdAt: Date = new Date(),
    updatedAt: Date = new Date(),
    completedAt: Date | null = null
  ) {
    this.id = id;
    this.userId = userId;
    this.title = title;
    this.description = description;
    this.status = status;
    this.priority = priority;
    this.dueDate = dueDate;
    this.tags = tags;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.completedAt = completedAt;
  }

  /**
   * 创建新任务
   */
  static create(
    userId: string,
    title: string,
    description: string,
    priority: Priority = Priorities.Medium
  ): Task {
    if (!userId || userId.trim().length === 0) {
      throw new Error('USER_ID_REQUIRED: 用户 ID 不能为空');
    }
    if (!title || title.trim().length === 0) {
      throw new Error('TASK_TITLE_EMPTY: 任务标题不能为空');
    }
    if (title.length > 200) {
      throw new Error('TASK_TITLE_TOO_LONG: 标题过长，最大 200 字符');
    }
    if (description.length > 5000) {
      throw new Error('TASK_DESCRIPTION_TOO_LONG: 描述过长，最大 5000 字符');
    }
    if (!isValidPriority(priority)) {
      throw new Error('INVALID_PRIORITY: 优先级无效');
    }

    const now = new Date();
    return new Task(
      randomUUID(),
      userId,
      title.trim(),
      description,
      priority,
      TaskStatuses.Pending,
      null,
      [],
      now,
      now,
      null
    );
  }

  /**
   * 更新任务信息
   */
  update(title?: string, description?: string, priority?: Priority): void {
    if (this.status === TaskStatuses.Completed) {
      throw new Error('TASK_ALREADY_COMPLETED: 已完成的任务不能更新');
    }

    if (title !== undefined) {
      if (!title || title.trim().length === 0) {
        throw new Error('TASK_TITLE_EMPTY: 任务标题不能为空');
      }
      if (title.length > 200) {
        throw new Error('TASK_TITLE_TOO_LONG: 标题过长，最大 200 字符');
      }
      this.title = title.trim();
    }

    if (description !== undefined) {
      if (description.length > 5000) {
        throw new Error('TASK_DESCRIPTION_TOO_LONG: 描述过长，最大 5000 字符');
      }
      this.description = description;
    }

    if (priority !== undefined) {
      if (!isValidPriority(priority)) {
        throw new Error('INVALID_PRIORITY: 优先级无效');
      }
      this.priority = priority;
    }

    this.updatedAt = new Date();
  }

  /**
   * 设置截止日期
   */
  setDueDate(dueDate: Date): void {
    if (dueDate < this.createdAt) {
      throw new Error('INVALID_DUE_DATE: 截止日期不能早于创建日期');
    }
    this.dueDate = dueDate;
    this.updatedAt = new Date();
  }

  /**
   * 标记为已完成
   */
  complete(): void {
    if (this.status === TaskStatuses.Completed) {
      throw new Error('TASK_ALREADY_COMPLETED: 任务已完成，不能再次完成');
    }
    this.status = TaskStatuses.Completed;
    this.completedAt = new Date();
    this.updatedAt = new Date();
  }

  /**
   * 添加标签
   */
  addTag(tag: Tag): void {
    if (!tag.name || tag.name.trim().length === 0) {
      throw new Error('TAG_NAME_EMPTY: 标签名不能为空');
    }
    if (this.tags.length >= 10) {
      throw new Error('TOO_MANY_TAGS: 标签过多，最多 10 个');
    }
    // 检查重复
    if (this.tags.some((t) => t.name === tag.name)) {
      throw new Error('DUPLICATE_TAG: 标签重复');
    }
    this.tags.push(tag);
    this.updatedAt = new Date();
  }

  /**
   * 移除标签
   */
  removeTag(tagName: string): void {
    this.tags = this.tags.filter((t) => t.name !== tagName);
    this.updatedAt = new Date();
  }
}

/**
 * 验证优先级是否有效
 */
function isValidPriority(priority: Priority): boolean {
  return (
    priority === Priorities.Low ||
    priority === Priorities.Medium ||
    priority === Priorities.High
  );
}

